proposals/records/005-sharding.md

```diff
@@ -0,0 +1,197 @@
+
+Responsibilities:
+- Know writer and reader health
+- Balancing shards across cluster
+    - Place new shards appropriately
+    - Move shards when things are unbalanced
+- Record keeping
+    - Shard stats
+    - Usage stats
+- Index queue consuming, operations
+  - Allow writing to multiple shards on a node at a time(right now, we only write to 1 at a time)
+
+APIs:
+- Create Shard
+- Delete Shard
+- Shard Operation
+    - Search
+    - etc
+
+
+#### Ingest changes
+
+Right now, ingest acts as our "Index Coordinator".
+
+The logic for how to store and query Index Nodes are spread throughout multiple components.
+
+This would be moved to the Index Coordinator.
+
+Changes to ingest/etc:
+- No longer responsible to directly connecting to node writer GPRC service to create new shards
+- No longer responsible for knowing what the best node to create a new shard on
+
+
+### Index Node Writer
+
+- sync changes of commit log to configured read replicas
+- shard create/delete is also part of commit log
+- configured read replicas for Index Node
+
+Write path diagram:
+![Write Path Diagram](./images/005/write-path-ndb-coor.drawio.png)
+
+
+#### Disks
+
+Use best performing persistent disks.
+
+Start with standard SSD persistent disks but move to extreme performance disks
+if performance required.
+
+### Index Node Read replica
+
+Read replicas are "dumb" and are mostly only responsible for serving
+requests from the replicated state.
+
+[OPTIONAL] Readers are responsible for merging segments
+
+Read path diagram:
+![Read Path Diagram](./images/005/read-path-ndb-coor.drawio.png)
+
+#### Disks
+
+To improve read replica performance, utilize local SSDs
+
+## Alternatives
+
+
+### Decouple writer and reader shards
+
+Allow shard writers and read replicas to be decoupled from one another.
+
+This would be accomplished through complex shard coordination schemes that
+allowed management of increasing particular KB shard replicas.
+
+
+#### Shard state coordination
+
+In order to manage consistency, we would need to use something like raft to
+coordinate reader replication state and have readers responsible for
+their own replication.
+
+The complexity around this is one of the reasons why this proposal focuses
+on having the writer responsible for replicating data.
+
+
+#### Not now
+
+This approach, while interesting, ends up with a lot of moving parts and becomes
+quite complex for little value.
+
+The approach outlined in this document allows us an easier path to iterate toward.
+
+
+## Rollout plan
+
+
+1. Unified index + commit log implementation
+2. Replicate commit log to configured read replicas
+3. Index coordinator to coordinate indexing to writers and query traffic to readers
+
+
+### Key Metrics
+
+- Index consumer latency/lag
+- Replication latency/lag
+- CPU/Disk on writers/readers
+- Disk IO performance
+
+## Future
+
+- Dynamically configure node read replicas
+- Consider moving to Kafka for a more performant ordered message queue
```

Comment:
Is Kafka better than NATS for this?

Reply:
Yes. We are currently using NATS in an awkward way in order to ensure message ordering: https://github.com/nuclia/nucliadb/blob/main/nucliadb/nucliadb/ingest/consumer/consumer.py#L106

something that kafka does by design

Reply:
It would be interesting to evaluate non-JVMs alternatives like Red Panda https://github.com/redpanda-data/redpanda/